### ISSUE TITLE
How to bootstrap Salt on RockyLinux?

### ISSUE THREAD

**Issue by maneamarius** [saltstack/salt-bootstrap#1803]
### Description of Issue/Question
I need to bootstrap Salt on RockyLinux, but I cannot get it working.
I need some help, please

### Setup
RockyLinux in Vagrant

### Steps to Reproduce Issue
1. downloaded bootstrap-salt.sh from develop branch
2. ran the command: `./bootstrap-salt.sh -x python3 -MPfq git develop'`

```
 *  WARN: Running the unstable version of bootstrap-salt.sh

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   4.18.0-348.2.1.el8_5.x86_64
 *  INFO:   Distribution: Rocky 8.5

 *  INFO: Installing minion
 *  INFO: Installing master
 *  INFO: Found function config_salt
 *  INFO: Found function preseed_master
 *  INFO: Found function daemons_running
 * ERROR: No dependencies installation function found. Exiting...
```

### Versions and Systems
```./bootstrap-salt.sh -v
./bootstrap-salt.sh -- Version 2021.09.17```

**Comment by myii** [comment]
@maneamarius The issue here is `git develop`, which would attempt to install from the old `develop` branch in the Salt repo:

* https://github.com/saltstack/salt/tree/develop

Use either of the following to get it working:

* `git master`
* `stable 3004` (or `stable latest`)

**Comment by maneamarius** [comment]
`./bootstrap-salt.sh git master`

`./bootstrap-salt.sh stable 3004`

`./bootstrap-salt.sh stable latest`

None of these worked, I am getting same error message..

``` 
 *  INFO: Running version: 2021.09.17
 *  INFO: Executed by: /bin/sh
 *  INFO: Command line: './bootstrap-salt.sh stable latest'
 *  WARN: Running the unstable version of bootstrap-salt.sh

 *  INFO: System Information:
 *  INFO:   CPU:          GenuineIntel
 *  INFO:   CPU Arch:     x86_64
 *  INFO:   OS Name:      Linux
 *  INFO:   OS Version:   4.18.0-348.2.1.el8_5.x86_64
 *  INFO:   Distribution: Rocky 8.5

 *  INFO: Installing minion
 *  INFO: Found function config_salt
 *  INFO: Found function preseed_master
 *  INFO: Found function daemons_running
 * ERROR: No dependencies installation function found. Exiting...
```

**Comment by myii** [comment]
@maneamarius I think I can see the problem here.  This is the relevant snippet from the output of the last `stable latest` test on Rocky Linux in this repo:

https://github.com/saltstack/salt-bootstrap/runs/5027699148?check_suite_focus=true#step:9:17

```console
*  INFO: Running version: 2021.09.17
*  INFO: Executed by: 
*  INFO: Command line: '/tmp/kitchen/bootstrap.sh -MP stable latest'
*  WARN: Running the unstable version of bootstrap-salt.sh

*  INFO: System Information:
*  INFO:   CPU:          GenuineIntel
*  INFO:   CPU Arch:     x86_64
*  INFO:   OS Name:      Linux
*  INFO:   OS Version:   5.11.0-1027-azure
*  INFO:   Distribution: CentOS 8.5

*  INFO: Installing minion
*  INFO: Installing master
*  INFO: Found function install_centos_stable_deps
*  INFO: Found function config_salt
*  INFO: Found function preseed_master
*  INFO: Found function install_centos_stable
*  INFO: Found function install_centos_stable_post
*  INFO: Found function install_centos_restart_daemons
*  INFO: Found function daemons_running
*  INFO: Found function install_centos_check_services
*  INFO: Running install_centos_stable_deps()
```

The key portion is that the Docker image is still being seen as `CentOS 8.5`, leading to the CentOS functions being found and used.  So in order to work on Vagrant, it looks like the bootstrap script needs to be updated to use these same functions when `Rocky 8.5` is detected.  So a block similar to the following, adjusted for Rocky Linux would probably do the trick:

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L4788-L4849

* The above won't be enough on its own, just giving an idea of what to look for.
* Searching through for Oracle Linux should highlight the other updates required as well.

**Comment by maneamarius** [comment]
sounds like your CI testing for Rocky is busted, as it `wrongly` passes.. :(
you could test using Vagrant box for Rocky, that is more reliable, I believe
EDIT: I can help with that if you'd like

**Comment by myii** [comment]
> sounds like your CI testing for Rocky is busted, as it `wrongly` passes.. :(

Not _your_ CI testing, I'm just another contributor here.

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/kitchen.yml#L103-L105

* https://hub.docker.com/r/rockylinux/rockylinux
  - Just noticed that this image hasn't been updated for around 3 months.

The CI isn't wrongly passing, since even if Rocky was detected properly, it would still be passed through to the CentOS functions to install Salt.  The Formulas' team is actually building images for Rocky, which are being used across numerous repos.

---

> you could test using Vagrant box for Rocky, that is more reliable, I believe

Vagrant testing in CI is a lot more expensive, though -- at the very least in terms of time and reliability (fails to even download boxes sometimes).  And with far more than 100 instances in the CI here, it simply isn't a viable option.

---

Digging deeper, I've tracked down the function that isn't working.

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L981-L990

(Not a problem) `lsb_release` isn't available, so skips further down.

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L1034

(Not a problem) Also not available.

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L1046-L1050

This results in all these files being detected:

```console
# cd /etc && /bin/ls *[_-]release *[_-]version 2>/dev/null | env -i sort | sed -e '/^redhat-release$/d' -e '/^lsb-release$/d'
centos-release
os-release
rocky-release
system-release
```

An important point to note here is that each of these files contain the right info:

```console
# cat /etc/centos-release
Rocky Linux release 8.5 (Green Obsidian)

# cat /etc/os-release
NAME="Rocky Linux"
VERSION="8.5 (Green Obsidian)"
ID="rocky"
ID_LIKE="rhel centos fedora"
VERSION_ID="8.5"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Rocky Linux 8.5 (Green Obsidian)"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:rocky:rocky:8.5:GA"
HOME_URL="https://rockylinux.org/"
BUG_REPORT_URL="https://bugs.rockylinux.org/"
ROCKY_SUPPORT_PRODUCT="Rocky Linux"
ROCKY_SUPPORT_PRODUCT_VERSION="8"

# cat /etc/rocky-release
Rocky Linux release 8.5 (Green Obsidian)

# cat /etc/system-release
Rocky Linux release 8.5 (Green Obsidian)
```

However, this info isn't used for the detection but rather the `shortname`:

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L1054-L1062

This is where the detection goes awry.  All because `centos-release` comes alphabetically before `rocky-release`.  It ends up resolving like this:

```console
n=centos
shortname=centos
rv=Rocky Linux release 8.5 (Green Obsidian)
v=8.5
```

Then, due to the `shortname`, gets selected by this entry in the `case` statement:

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L1077

---

> EDIT: I can help with that if you'd like

You'll probably have to, there may not be the required bandwidth to resolve this.

---

I've searched further on Docker Hub and there's now an official image:

* https://hub.docker.com/_/rockylinux?tab=tags
  - Like above, hasn't been updated for 2 months.

Unfortunately, this image suffers from the same problem above.

**Comment by myii** [comment]
I've been able to reproduce this bug on AlmaLinux as well.  These are the key lines involved:

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L945-L948

```diff
-    KNOWN_RELEASE_FILES=$(echo "(arch|alpine|centos|debian|ubuntu|fedora|redhat|suse|\
+    KNOWN_RELEASE_FILES=$(echo "(arch|alpine|almalinux|centos|debian|ubuntu|fedora|redhat|suse|\
```

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L961-L962

```diff
-    max_prio="redhat-release centos-release oracle-release fedora-release"
+    max_prio="redhat-release centos-release oracle-release fedora-release almalinux-release"
```

https://github.com/saltstack/salt-bootstrap/blob/e4855afb3c678912329a40d4685166943a210d07/bootstrap-salt.sh#L1076-L1077

```diff
             arch               ) n="Arch Linux"     ;;
+            almalinux          ) n="AlmaLinux"      ;;
             alpine             ) n="Alpine Linux"   ;;
```

---

The good news is that this isn't true:

> This is where the detection goes awry. All because `centos-release` comes alphabetically before `rocky-release`.

Making the adjustment to the `max_prio` variable does the job.  So an overall fix for both Rocky Linux and AlmaLinux shouldn't be so difficult, after all.

**Comment by myii** [comment]
@maneamarius I've proposed a PR to fix this issue (#1804).  You can try it out by downloading the bootstrap script from that PR's branch:

* https://raw.githubusercontent.com/myii/salt-bootstrap/feat/fix-1803_add-detection-for-almalinux-and-rocky/bootstrap-salt.sh

I'd appreciate it if you could test it out and add a comment to the PR, to mention if it works for you or not.

**Comment by maneamarius** [comment]
Very good analysis, thanks for taking the time to check it!
I'll test with Rocky and Alma and update the PR.
Sorry for the previous assumption, I wrongly assumed you are the owner of this project :))
